proctor-common/src/main/java/com/indeed/proctor/common/el/PartialExpressionBuilder.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.indeed.proctor.common.el;
+
+import org.apache.el.MethodExpressionImpl;
+import org.apache.el.MethodExpressionLiteral;
+import org.apache.el.ValueExpressionImpl;
+import org.apache.el.lang.FunctionMapperFactory;
+import org.apache.el.lang.VariableMapperFactory;
+import org.apache.el.parser.AstDeferredExpression;
+import org.apache.el.parser.AstDynamicExpression;
+import org.apache.el.parser.AstFunction;
+import org.apache.el.parser.AstIdentifier;
+import org.apache.el.parser.AstLiteralExpression;
+import org.apache.el.parser.AstValue;
+import org.apache.el.parser.ELParser;
+import org.apache.el.parser.Node;
+import org.apache.el.parser.NodeVisitor;
+import org.apache.el.parser.ParseException;
+import org.apache.el.util.MessageFactory;
+
+import javax.el.ELContext;
+import javax.el.ELException;
+import javax.el.FunctionMapper;
+import javax.el.MethodExpression;
+import javax.el.ValueExpression;
+import javax.el.VariableMapper;
+
+import java.io.StringReader;
+import java.lang.reflect.Method;
+import java.util.Set;
+
+/**
+ * Like ExpressionBuilder except removes nodes referring to missing variables
```

Comment:
I am not sure what the use case is for this. Removing missing variables just hides bugs in the users set up. They should be passing in the needed variables of the rules or else you will not be correctly allocating based on the rules.

Can you provide more information on what this will be used for?

Reply:
Yes it's for searching for allocations matching part of a context. E. g. if I want to find all allocations that might apply to country US or a particular surface id etc.

Reply:
Gotcha, just wanted to confirm justification of this existing in proctor vs your tool

Reply:
Yeah with the refactor we could conceivably put this in a separate project if we made the RuleEvaluator constructor public, but I think this functionality is very useful in general and a bit difficult to implement. We've had many cases where this sort of thing would have been a lifesaver.